src/resmom/mom_pmix.c

```diff
@@ -0,0 +1,1537 @@
+	PBS_PMIX_OPER_UNDEFINED
+} pbs_pmix_oper_type_t;
+
+/* Data structure to house auxiliary PMIx operation data */
+typedef struct pbs_pmix_oper {
+	pbs_pmix_oper_type_t type;
+	job *pjob;
+	struct pbs_pmix_oper *next;
+	union {
+		pbs_pmix_fence_data_t fence;
+		pbs_pmix_spawn_data_t spawn;
+	} data;
+} pbs_pmix_oper_t;
+
+#endif
+
+/* Locking structure and macros */
+typedef struct {
+	pthread_mutex_t mutex;
+	pthread_cond_t cond;
+	volatile bool active;
+	pmix_status_t status;
+} pbs_pmix_lock_t;
+
+#define PBS_PMIX_CONSTRUCT_LOCK(l) \
+	do { \
+		pthread_mutex_init(&(l)->mutex, NULL); \
+		pthread_cond_init(&(l)->cond, NULL); \
+		(l)->active = true; \
+		(l)->status = PMIX_SUCCESS; \
+	} while(0)
+
+#define PBS_PMIX_DESTRUCT_LOCK(l) \
+	do { \
+		pthread_mutex_destroy(&(l)->mutex); \
+		pthread_cond_destroy(&(l)->cond); \
+	} while(0)
+
+#define PBS_PMIX_WAIT_THREAD(lck) \
+	do { \
+		pthread_mutex_lock(&(lck)->mutex); \
+		while ((lck)->active) { \
+			pthread_cond_wait(&(lck)->cond, &(lck)->mutex); \
+		} \
+		pthread_mutex_unlock(&(lck)->mutex); \
+	} while(0)
+
+#define PBS_PMIX_WAKEUP_THREAD(lck) \
+	do { \
+		pthread_mutex_lock(&(lck)->mutex); \
+		(lck)->active = false; \
+		pthread_cond_broadcast(&(lck)->cond); \
+		pthread_mutex_unlock(&(lck)->mutex); \
+	} while(0)
+
+/**
+ * @brief
+ * This callback function is invoked by the PMIx library after it
+ * has been notified a process has exited.
+ *
+ * @param[in] status - exit status of the process
+ * @param[in] cbdata - opaque callback data passed to the caller
+ *
+ * @retval void
+ *
+ * @note
+ * This function may be superfluous, in which case the call to
+ * PMIx_Notify_event() should be passed NULL in pbs_pmix_notify_exit()
+ * for its callback and this funtion removed. It has been left in for
+ * the time being so that the log shows it being called.
+ */
+static void
+pbs_pmix_notify_exit_cb(pmix_status_t status, void *cbdata)
+{
+	log_event(PBSEVENT_DEBUG3, 0, LOG_DEBUG, __func__, "called");
+	log_event(PBSEVENT_DEBUG3, 0, LOG_DEBUG, __func__, "returning");
+}
+
+/**
+ * @brief
+ * Notify PMIx that a task has exited by constructing a PMIx info
+ * array and passing it to PMIx_Notify_event.
+ *
+ * @param[in] pjob - pointer to the job structure for the exited task
+ * @param[in] exitstat - numeric exit status of the task
+ * @param[in] msg - optional message supplied by the caller
+ *
+ * @retval void
+ */
+void
+pbs_pmix_notify_exit(job *pjob, int exitstat, char *msg)
+{
+	pmix_status_t status;
+	pmix_info_t *pinfo;
+	size_t ninfo;
+	pmix_proc_t procname, procsrc;
+	bool flag = true;
+
+	log_event(PBSEVENT_DEBUG3, 0, LOG_DEBUG, __func__, "called");
+	if (!pjob) {
+		log_event(PBSEVENT_DEBUG, 0, LOG_ERR, __func__,
+			"No job supplied, returning");
+		return;
+	}
+	snprintf(log_buffer, sizeof(log_buffer),
+		"Setting up the info array for termination");
+	log_event(PBSEVENT_JOB, PBS_EVENTCLASS_JOB, LOG_INFO,
+		pjob->ji_qs.ji_jobid, log_buffer);
+	ninfo = 3;
```

Comment:
shall we add a comment to explain why we need ninfo to be 3 here.

Reply:
could you please reply to or address this comment.

Reply:
If the comment added is insufficient, could you please indicate the information you would like it to convey? We are creating an info array for three or four items, so I'm not sure what else to say.